sw/device/lib/dif/dif_usbdev.h

```diff
@@ -0,0 +1,759 @@
+ * remove. `top` points to the last free buffer added to the pool. The pool is
+ * full when `top == USBEV_NUM_BUFFERS - 1` and empty when `top == -1`.
+ */
+typedef struct dif_usbdev_buffer_pool {
+  uint8_t entries[USBDEV_NUM_BUFFERS];
+  int8_t top;
+} dif_usbdev_buffer_pool_t;
+
+/**
+ * State of a read buffer operation.
+ *
+ * This struct stores
+ * - the buffer to read from,
+ * - the number of remaining bytes, and
+ * - the offset to start reading from
+ * to track the state of a read buffer operation internally. This allows clients
+ * to freely choose their buffer sizes and easily process packets that are
+ * larger than their buffers by simply calling `dif_usbdev_rx_packet_read_bytes`
+ * repeatedly until they read the entire payload of each packet.
+ *
+ * See also: `dif_usbdev_rx_packet_get_info`, `dif_usbdev_rx_packet_read_bytes`.
+ */
+typedef struct dif_usbdev_read_buffer_op_state {
+  uint32_t buffer_entry;
+  uint32_t remaining_bytes;
+  uint32_t offset;
+} dif_usbdev_read_buffer_op_state_t;
+
+/**
+ * Internal state of a USB device.
+ *
+ * Instances of this struct must be initialized by `dif_usbdev_init()` before
+ * being passed to other functions in this library. Its fields should be
+ * considered private and are only provided here so that callers can allocate
+ * it.
+ */
+typedef struct dif_usbdev {
+  mmio_region_t base_addr;
+  dif_usbdev_buffer_pool_t buffer_pool;
+  dif_usbdev_read_buffer_op_state_t read_buffer_op_state;
+  /**
+   * Active buffer operations.
+   *
+   * Bits 0..11: True if writing bytes of an outgoing packet for endpoint N
+   * (0..11).
+   * Bits 12..30: Unused.
+   * Bit 31: True if reading bytes of an incoming packet.
+   */
+  uint32_t active_buffer_ops;
+} dif_usbdev_t;
+
+/**
+ * Enumeration for enabling/disabling various functionality.
+ */
+typedef enum dif_usbdev_toggle {
+  kDifUsbdevToggleDisable,
+  kDifUsbdevToggleEnable,
+} dif_usbdev_toggle_t;
+
+/**
+ * Set of allowed configurations for USB power sense override.
+ */
+typedef enum dif_usbdev_power_sense_override {
+  kDifUsbdevPowerSenseOverrideDisabled,
+  kDifUsbdevPowerSenseOverridePresent,
+  kDifUsbdevPowerSenseOverrideNotPresent
+} dif_usbdev_power_sense_override_t;
+
+/**
+ * Configuration for initializing a USB device.
+ */
+typedef struct dif_usbdev_config {
+  mmio_region_t base_addr;
+  dif_usbdev_toggle_t differential_rx;
+  dif_usbdev_toggle_t differential_tx;
+  dif_usbdev_toggle_t single_bit_eop;
+  dif_usbdev_power_sense_override_t power_sense_override;
+} dif_usbdev_config_t;
+
+/**
+ * Common return codes for the functions in this library.
+ *
+ * `kDifUsbdevResultInvalidArgument` signals that there was a problem with the
+ * argument(s) but the function did not cause any side effects.
+ * `kDifUsbdevResultGeneralError` covers other error cases.
+ */
+typedef enum dif_usbdev_result {
+  kDifUsbdevResultOK,
+  kDifUsbdevResultInvalidArgument,
```

Comment:
Minor nit: `OK` -> `Ok`, and `InvalidArgument` -> `BadArg` for consistency with other DIFs.